src/Core/Sql.php

```diff
@@ -0,0 +1,177 @@
+<?php
+
+namespace FlexForm\Core;
+
+use FlexForm\FlexFormException;
+use MediaWiki\MediaWikiServices;
+use MediaWiki\Revision\RevisionRecord;
+use MediaWiki\Storage\EditResult;
+use MediaWiki\User\UserIdentity;
+use WikiPage;
+
+class Sql {
+
+	private const DBTABLE = 'FLEXFORM';
+
+	/**
+	 * @param $updater
+	 *
+	 * @return bool
+	 * @throws \MWException
+	 */
+	public static function addTables( $updater ) {
+		$dbt = $updater->getDB()->getType();
+		// If using SQLite, just use the MySQL/MariaDB schema, it's compatible
+		// anyway. Only PGSQL and some more exotic variants need a totally
+		// different schema.
+		if ( $dbt === 'sqlite' ) {
+			$dbt = 'sql';
+		}
+		$tables = __DIR__ . "/../../sql/FlexForm.$dbt";
+
+		if ( file_exists( $tables ) ) {
+			$updater->addExtensionUpdate( array(
+											  'addTable',
+											  self::DBTABLE,
+											  $tables,
+											  true
+										  ) );
+		} else {
+			throw new \MWException(
+				wfMessage(
+					'flexform-unsupported-database',
+					$dbt
+				)
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * @param WikiPage $article
+	 * @param UserIdentity $user
+	 * @param string $summary
+	 * @param int $flags
+	 * @param RevisionRecord $revisionRecord
+	 * @param EditResult $editResult
+	 *
+	 * @return bool
+	 * @throws Exception
+	 * @throws FlexFormException
+	 */
+	public static function pageSaved(
+		WikiPage $article,
+		UserIdentity $user,
+		string $summary,
+		int $flags,
+		RevisionRecord $revisionRecord,
+		EditResult $editResult
+	) : bool {
+		$id = $article->getId();
+		$idExists = self::exists( $id );
+		if ( Rights::isUserAllowedToEditorCreateForms() ) {
+			if ( $idExists ) {
+				return true;
+			} else {
+				$result = self::addPageId( $id );
+				if ( $result === false ) {
+					throw new FlexFormException( 'Can\'t save to Database [add]' );
+				}
+			}
+		} else {
+			if ( $idExists ) {
+				$result = self::removePageId( $id );
```

Comment:
idempotent: no matter how many times you run the code, the result will always be the same... You are referring to the check if the ID exists beforehand I guess?
In this case I am saving a DB action, since I already know if the ID exists.

Reply:
Not really, since if you remove the `if ( $idExists )` statement here, the ID check can be moved into the consequent block of the `if`-statement, also saving one database call.

Reply:
But didn't I already do this ?